In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot

In [2]:
df=pd.read_csv('data.csv')

In [3]:
df

Suburb              Address  Rooms Type      Price Method  \
0          Abbotsford        68 Studley St    2.0    h        NaN     SS   
1          Abbotsford         85 Turner St    2.0    h  1480000.0      S   
2          Abbotsford      25 Bloomburg St    2.0    h  1035000.0      S   
3          Abbotsford   18/659 Victoria St    3.0    u        NaN     VB   
4          Abbotsford         5 Charles St    3.0    h  1465000.0     SP   
5          Abbotsford     40 Federation La    3.0    h   850000.0     PI   
6          Abbotsford          55a Park St    4.0    h  1600000.0     VB   
7          Abbotsford         16 Maugie St    4.0    h        NaN     SN   
8          Abbotsford         53 Turner St    2.0    h        NaN      S   
9          Abbotsford         99 Turner St    2.0    h        NaN      S   
10         Abbotsford       129 Charles St    2.0    h   941000.0      S   
11         Abbotsford         124 Yarra St    3.0    h  1876000.0      S   
12         Abbotsford  121/56 Nicholson St    2.0    u        NaN     PI   
13         Abbotsford        17 Raphael St    4.0    h        NaN      W   
14         Abbotsford        98 Charles St    2.0    h  1636000.0      S   
15         Abbotsford     217 Langridge St    3.0    h  1000000.0      S   
16         Abbotsford      18a Mollison St    2.0    t   745000.0      S   
17         Abbotsford   6/241 Nicholson St    1.0    u   300000.0      S   
18         Abbotsford        10 Valiant St    2.0    h  1097000.0      S   
19         Abbotsford  403/609 Victoria St    2.0    u   542000.0      S   
20         Abbotsford            2 Rich St    2.0    h        NaN     SP   
21         Abbotsford    25/84 Trenerry Cr    2.0    u   760000.0     SP   
22         Abbotsford    106/119 Turner St    1.0    u   481000.0     SP   
23         Abbotsford   411/8 Grosvenor St    2.0    u   700000.0     VB   
24         Abbotsford      40 Nicholson St    3.0    h  1350000.0     VB   
25         Abbotsford  123/56 Nicholson St    2.0    u   750000.0      S   
26         Abbotsford           22 Park St    4.0    h  1985000.0      S   
27         Abbotsford    13/84 Trenerry Cr    1.0    u   500000.0      S   
28         Abbotsford        45 William St    2.0    h  1172500.0      S   
29         Abbotsford   7/20 Abbotsford St    1.0    u   441000.0     SP   
...               ...                  ...    ...  ...        ...    ...   
23648   Vermont South        2 Settlers Ct    4.0    h  1295000.0     SP   
23649   Vermont South        9 Winswood Cl    4.0    h  1250000.0     PI   
23650        Viewbank        149 Graham Rd    5.0    h  1316000.0     SP   
23651        Viewbank       7 Northwood Dr    4.0    h        NaN     SP   
23652        Wantirna   16 chesterfield Ct    4.0    h   951000.0      S   
23653        Wantirna            9 eton Sq    4.0    h        NaN     PI   
23654  Wantirna South         12 Armagh Cr    4.0    h  1323000.0      S   
23655  Wantirna South        34 Fewster Dr    3.0    h   970000.0      S   
23656  Wantirna South           15 Mara Cl    4.0    h  1330000.0      S   
23657        Watsonia       2/14 Daours Ct    2.0    u   590000.0      S   
23658        Watsonia        76 Kenmare St    2.0    h   650000.0     PI   
23659  Watsonia North         22 Meakin St    3.0    h        NaN     PI   
23660        Werribee          6 Feijoa Ct    5.0    h        NaN      W   
23661        Werribee       11 Gundaroo Sq    4.0    h        NaN     PI   
23662        Werribee         91 Latham St    4.0    h   540000.0     PI   
23663        Werribee          5 Nuragi Ct    4.0    h   635000.0      S   
23664        Werribee             3 Sky Ct    3.0    h        NaN     PI   
23665  West Footscray      1/21 Waratah St    2.0    t   655000.0      S   
23666     Westmeadows           9 Black St    3.0    h   582000.0      S   
23667     Westmeadows      81 Hillcrest Dr    4.0    h   720000.0     VB   
23668   Wheelers Hill         12 Strada Cr    4.0    h  1245000

1- Determine the no. of rows and no. of columns in the dataset?

In [4]:
df.shape

(23678, 21)

              
              The data have 23678 rows and 21 columns.

2. Display the datatypes of each column. 

In [5]:
df.dtypes

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

3. Determine the row indexes of rows having all null values (i.e. empty rows).

In [6]:
df.isnull().sum()

Suburb             131
Address            131
Rooms              131
Type               131
Price             5282
Method             131
SellerG            131
Date               131
Distance           132
Postcode           132
Bedroom2          4612
Bathroom          4615
Car               4757
Landsize          6268
BuildingArea     13660
YearBuilt        12138
CouncilArea       8022
Lattitude         4435
Longtitude        4435
Regionname         132
Propertycount      132
dtype: int64

4. Determine the total number of empty rows and their percentage in whole dataset and remove them. 

In [12]:
df.isnull().describe()

Suburb Address  Rooms   Type  Price Method SellerG   Date Distance  \
count   23678   23678  23678  23678  23678  23678   23678  23678    23678   
unique      2       2      2      2      2      2       2      2        2   
top     False   False  False  False  False  False   False  False    False   
freq    23547   23547  23547  23547  18396  23547   23547  23547    23546   

       Postcode  ... Bathroom    Car Landsize BuildingArea YearBuilt  \
count     23678  ...    23678  23678    23678        23678     23678   
unique        2  ...        2      2        2            2         2   
top       False  ...    False  False    False         True      True   
freq      23546  ...    19063  18921    17410        13660     12138   

       CouncilArea Lattitude Longtitude Regionname Propertycount  
count        23678     23678      23678      23678         23678  
unique           2         2          2          2             2  
top          False     False      False      False         False  
freq         15656     19243      19243      23546         23546  

[4 rows x 21 columns]

5. Determine the column names having missing values.

In [9]:
df.columns[df.isnull().any()]

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

6. Determine the missing values and their percentage in each column.

In [13]:

df.isnull().sum() * 100 / len(df)

Suburb            0.553256
Address           0.553256
Rooms             0.553256
Type              0.553256
Price            22.307627
Method            0.553256
SellerG           0.553256
Date              0.553256
Distance          0.557480
Postcode          0.557480
Bedroom2         19.477996
Bathroom         19.490666
Car              20.090379
Landsize         26.471830
BuildingArea     57.690683
YearBuilt        51.262776
CouncilArea      33.879551
Lattitude        18.730467
Longtitude       18.730467
Regionname        0.557480
Propertycount     0.557480
dtype: float64

7. Determine the row indexes having missing values. 

In [17]:
df[df.isnull().any(axis=1)]

Suburb              Address  Rooms Type      Price Method  \
0          Abbotsford        68 Studley St    2.0    h        NaN     SS   
1          Abbotsford         85 Turner St    2.0    h  1480000.0      S   
3          Abbotsford   18/659 Victoria St    3.0    u        NaN     VB   
5          Abbotsford     40 Federation La    3.0    h   850000.0     PI   
7          Abbotsford         16 Maugie St    4.0    h        NaN     SN   
8          Abbotsford         53 Turner St    2.0    h        NaN      S   
9          Abbotsford         99 Turner St    2.0    h        NaN      S   
10         Abbotsford       129 Charles St    2.0    h   941000.0      S   
12         Abbotsford  121/56 Nicholson St    2.0    u        NaN     PI   
13         Abbotsford        17 Raphael St    4.0    h        NaN      W   
15         Abbotsford     217 Langridge St    3.0    h  1000000.0      S   
16         Abbotsford      18a Mollison St    2.0    t   745000.0      S   
17         Abbotsford   6/241 Nicholson St    1.0    u   300000.0      S   
19         Abbotsford  403/609 Victoria St    2.0    u   542000.0      S   
20         Abbotsford            2 Rich St    2.0    h        NaN     SP   
21         Abbotsford    25/84 Trenerry Cr    2.0    u   760000.0     SP   
22         Abbotsford    106/119 Turner St    1.0    u   481000.0     SP   
23         Abbotsford   411/8 Grosvenor St    2.0    u   700000.0     VB   
26         Abbotsford           22 Park St    4.0    h  1985000.0      S   
27         Abbotsford    13/84 Trenerry Cr    1.0    u   500000.0      S   
28         Abbotsford        45 William St    2.0    h  1172500.0      S   
29         Abbotsford   7/20 Abbotsford St    1.0    u   441000.0     SP   
31                NaN                  NaN    NaN  NaN        NaN    NaN   
32         Abbotsford   16/11 Nicholson St    3.0    u        NaN     PI   
34         Abbotsford     250 Langridge St    2.0    t   847000.0      S   
35         Abbotsford      16b Mollison St    2.0    h        NaN     PI   
37         Abbotsford     196 Nicholson St    3.0    h   955000.0      S   
40         Abbotsford      1/1 Mayfield St    3.0    h        NaN     SP   
41         Abbotsford   13/11 Nicholson St    3.0    t   900000.0      S   
42         Abbotsford  138/56 Nicholson St    3.0    u  1090000.0      S   
...               ...                  ...    ...  ...        ...    ...   
23648   Vermont South        2 Settlers Ct    4.0    h  1295000.0     SP   
23649   Vermont South        9 Winswood Cl    4.0    h  1250000.0     PI   
23650        Viewbank        149 Graham Rd    5.0    h  1316000.0     SP   
23651        Viewbank       7 Northwood Dr    4.0    h        NaN     SP   
23652        Wantirna   16 chesterfield Ct    4.0    h   951000.0      S   
23653        Wantirna            9 eton Sq    4.0    h        NaN     PI   
23654  Wantirna South         12 Armagh Cr    4.0    h  1323000.0      S   
23655  Wantirna South        34 Fewster Dr    3.0    h   970000.0      S   
23656  Wantirna South           15 Mara Cl    4.0    h  1330000.0      S   
23657        Watsonia       2/14 Daours Ct    2.0    u   590000.0      S   
23658        Watsonia        76 Kenmare St    2.0    h   650000.0     PI   
23659  Watsonia North         22 Meakin St    3.0    h        NaN     PI   
23660        Werribee          6 Feijoa Ct    5.0    h        NaN      W   
23661        Werribee       11 Gundaroo Sq    4.0    h        NaN     PI   
23662        Werribee         91 Latham St    4.0    h   540000.0     PI   
23663        Werribee          5 Nuragi Ct    4.0    h   635000.0      S   
23664        Werribee             3 Sky Ct    3.0    h        NaN     PI   
23665  West Footscray      1/21 Waratah St    2.0    t   655000.0      S   
23666     Westmeadows           9 Black St    3.0    h   582000.0      S   
23667     Westmeadows      81 Hillcrest Dr    4.0    h   720000.0     VB   
23668   Wheelers Hill         12 Strada Cr    4.0    h  1245000

8. Determine the missing values and their percentage in each row.  

In [18]:
df.apply(lambda x: x.count(), axis=1)

0        18
1        19
2        21
3        18
4        21
5        19
6        21
7        20
8        19
9        19
10       19
11       21
12       20
13       20
14       21
15       12
16       12
17       19
18       21
19       12
20       20
21       12
22       12
23       19
24       21
25       21
26       14
27       12
28       19
29       19
         ..
23648    19
23649    18
23650    18
23651    19
23652    20
23653    17
23654    18
23655    18
23656    20
23657    19
23658    20
23659    16
23660    17
23661    16
23662    12
23663    20
23664    16
23665    19
23666    18
23667    17
23668    19
23669    20
23670    19
23671    19
23672    20
23673    19
23674    13
23675    17
23676    20
23677    19
Length: 23678, dtype: int64

9. Determine the column names having more than 30% missing values and remove them from the dataset.

In [45]:
limitPer = len(df) * .30
df = df.dropna(thresh=limitPer,axis=1)

10. Determine the number and percentage of rows having more than 5 missing values. If the percentage of such 
rows is less than 20% then remove all these rows. 


In [48]:
df.drop(df[df.isnull().sum(axis=1)>5].index,axis=0,inplace=True)

print(round(100*(1-df.count()/len(df)),2))

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.67
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          0.00
Bathroom          0.01
Car               0.70
Landsize          8.64
BuildingArea     47.43
YearBuilt        39.44
CouncilArea      17.84
Lattitude         0.14
Longtitude        0.14
Regionname        0.00
Propertycount     0.00
dtype: float64


11. Display the description of data and percentage of missing values still left column wise. 

In [51]:
perc =[.20, .40, .60, .80]
include =['object', 'float', 'int']
df.describe(percentiles = perc, include = include)

Suburb       Address         Rooms   Type         Price Method  \
count       19055         19055  19055.000000  19055  1.492600e+04  19055   
unique        322         18696           NaN      3           NaN      9   
top     Reservoir  5 Charles St           NaN      h           NaN      S   
freq          450             5           NaN  13326           NaN  11221   
mean          NaN           NaN      2.975020    NaN  1.067117e+06    NaN   
std           NaN           NaN      0.972329    NaN  6.382375e+05    NaN   
min           NaN           NaN      1.000000    NaN  8.500000e+04    NaN   
20%           NaN           NaN      2.000000    NaN  5.950000e+05    NaN   
40%           NaN           NaN      3.000000    NaN  7.920000e+05    NaN   
50%           NaN           NaN      3.000000    NaN  9.000000e+05    NaN   
60%           NaN           NaN      3.000000    NaN  1.024000e+06    NaN   
80%           NaN           NaN      4.000000    NaN  1.430000e+06    NaN   
max           NaN           NaN     10.000000    NaN  9.000000e+06    NaN   

       SellerG        Date      Distance      Postcode  ...      Bathroom  \
count    19055       19055  19055.000000  19055.000000  ...  19054.000000   
unique     309          58           NaN           NaN  ...           NaN   
top     Nelson  23-09-2017           NaN           NaN  ...           NaN   
freq      2002         662           NaN           NaN  ...           NaN   
mean       NaN         NaN     10.222556   3109.772763  ...      1.570956   
std        NaN         NaN      6.055921     93.808492  ...      0.712765   
min        NaN         NaN      0.000000   3000.000000  ...      0.000000   
20%        NaN         NaN      5.300000   3041.000000  ...      1.000000   
40%        NaN         NaN      8.000000   3073.000000  ...      1.000000   
50%        NaN         NaN      9.200000   3101.000000  ...      1.000000   
60%        NaN         NaN     11.000000   3124.000000  ...      2.000000   
80%        NaN         NaN     13.900000   3165.000000  ...      2.000000   
max        NaN         NaN     48.100000   3977.000000  ...     12.000000   

                 Car       Landsize  BuildingArea     YearBuilt  CouncilArea  \
count   18921.000000   17408.000000  10018.000000  11540.000000        15656   
unique           NaN            NaN           NaN           NaN           34   
top              NaN            NaN           NaN           NaN   Boroondara   
freq             NaN            NaN           NaN           NaN         1677   
mean        1.626235     551.822151    154.527895   1964.636742          NaN   
std         0.974048    3544.489659    462.535765     37.595504          NaN   
min         0.000000       0.000000      0.000000   1196.000000          NaN   
20%         1.000000     140.000000     86.000000   1930.000000          NaN   
40%         1.000000     318.000000    115.000000   1960.000000          NaN   
50%         2.000000     448.000000    129.000000   1970.000000          NaN   
60%         2.000000     560.000000    145.000000   1975.000000          NaN   
80%         2.000000     696.000000    196.000000   2002.200000          NaN   
max        26.000000  433014.000000  44515.000000   2106.000000          NaN   

           Lattitude    Longtitude             Regionname Propertycount  
count   19029.000000  19029.000000                  19055  19055.000000  
unique           NaN           NaN                      8           NaN  
top              NaN           NaN  Southern Metropolitan           NaN  
freq             NaN           NaN                   7032           NaN  
mean      -37.812187    145.000471                    NaN   7474.095565  
std         0.080002      0.106099                    NaN   4312.653403  
min       -38.184180    144.431620                    NaN    129.000000  
20%       -37.872224    144.918880                    NaN   3619.000000  
40%       -37.829836    144.987700                    NaN   5549.0

12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? 
Give a brief explanation and treat the missing values in 'Price' column. 


In [52]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St    2.0    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St    3.0    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [54]:
df['Price'].isnull().sum()

4129

If i have to train the data as a target with price column i can replace price columns with its median or mean for entire dataset for best prediction result.

In [55]:
df['Price'].fillna((df['Price'].mean()), inplace=True)

In [56]:
df['Price'].isnull().sum()

0

   Now there is no null values present in df['price'] ,  all the values are replace by its mean value.

13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most 
suitable method.


In [57]:

df.isnull().sum() * 100 / len(df)

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price             0.000000
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.000000
Postcode          0.000000
Bedroom2          0.000000
Bathroom          0.005248
Car               0.703227
Landsize          8.643401
BuildingArea     47.425872
YearBuilt        39.438468
CouncilArea      17.837838
Lattitude         0.136447
Longtitude        0.136447
Regionname        0.000000
Propertycount     0.000000
dtype: float64

In [59]:
df=df.fillna(df.mean())

In [60]:
df.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
YearBuilt           0
CouncilArea      3399
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

In [63]:

df.drop(['CouncilArea'], axis = 1,inplace = True)

In [64]:
df.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

All missing values are treated.

14. Now since you have completed your data cleaning, determine the number and percentage of columns removed.

In [70]:
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude',
       'Regionname', 'Propertycount'],
      dtype='object')

In [74]:
df['percent'] = (df['Price'] / df['Price'].sum()) * 100

In [76]:
df.head()

Suburb             Address  Rooms Type         Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0    h  1.067117e+06     SS  Jellis   
1  Abbotsford        85 Turner St    2.0    h  1.480000e+06      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0    h  1.035000e+06      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0    u  1.067117e+06     VB  Rounds   
4  Abbotsford        5 Charles St    3.0    h  1.465000e+06     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0    154.527895   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0    154.527895   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0     79.000000   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0    154.527895   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0    150.000000   

     YearBuilt  Lattitude  Longtitude             Regionname Propertycount  \
0  1964.636742   -37.8014    144.9958  Northern Metropolitan        4019.0   
1  1964.636742   -37.7996    144.9984  Northern Metropolitan        4019.0   
2  1900.000000   -37.8079    144.9934  Northern Metropolitan        4019.0   
3  1964.636742   -37.8114    145.0116  Northern Metropolitan        4019.0   
4  1900.000000   -37.8093    144.9944  Northern Metropolitan        4019.0   

    percent  
0  0.005248  
1  0.007278  
2  0.005090  
3  0.005248  
4  0.007205  

[5 rows x 21 columns]

15. Determine the number and percentage of rows removed in whole process. Also comment briefly that according 
to you the amount of data lost is justified or not. 


In [77]:
df.shape

(19055, 21)

Now our dataset have 19055 rows which means 4622 rows are deleted and remaing are replaced , so i think this is best for the model because by so much drop data the model does not learn so much or by access null values model does not work , by which it goes into overfitting and underfitting easily.